cmd/hub/mcscontroller/controller.go

```diff
@@ -117,16 +124,138 @@ func (r *ServiceExportReconciler) Reconcile(ctx context.Context, req ctrl.Reques
+				r.Client.Delete(ctx, &si)
+			} else {
+				if !apierrors.IsNotFound(err) {
+					log.Error(err, "Failed to get the service Import to delete")
+					return ctrl.Result{}, err
+				}
+			}
+		} else {
+			lhutil.UpdateServiceExportConditions(ctx, &se, log, mcsv1a1.ServiceExportConflict, corev1.ConditionFalse, "", "successfully update the service export")
+			err = r.Client.Status().Update(ctx, &se)
+			if err != nil {
+				return ctrl.Result{}, err
+			}
+			r.ensureImportFor(ctx, &se)
+		}
+	}
 	return ctrl.Result{}, nil
 }
 
+// getPrimaryExportObject finds the Primary ExportSpec from ServiceExportList.
+// Retruns the ExportSpec object that determined as the primary one from
+// the given ServiceExportList according to the conflict resolution specification set in the
+// KEP, that is implemented in IsPreferredOver method of ExportSpec.
+func getPrimaryExportObject(exportList mcsv1a1.ServiceExportList) (*mcs.ExportSpec, error) {
+	var primary *mcs.ExportSpec
+	for i, se := range exportList.Items {
+		exportSpec := &mcs.ExportSpec{}
+		err := exportSpec.UnmarshalObjectMeta(&se.ObjectMeta)
+		if err != nil {
+			return nil, err
+		}
+		if i == 0 || exportSpec.IsPreferredOver(primary) {
+			primary = exportSpec
+		}
+	}
+	return primary, nil
+}
+
+// create or update the ServiceImport corresponding to the provided export.
+func (r *ServiceExportReconciler) ensureImportFor(ctx context.Context, se *mcsv1a1.ServiceExport) error {
+	es := &mcs.ExportSpec{}
+
+	err := es.UnmarshalObjectMeta(&se.ObjectMeta)
+	if err != nil {
+		return err
+	}
+
+	namespacedName := types.NamespacedName{
+		Namespace: se.Name,
+		Name:      se.Namespace,
+	}
+	log := r.Log.WithValues("service", namespacedName.Name)
+
+	si := mcsv1a1.ServiceImport{}
+	err = r.Client.Get(ctx, namespacedName, &si)
+	expected := mcsv1a1.ServiceImport{
+		ObjectMeta: metav1.ObjectMeta{
+			Namespace: es.Namespace,
+			Name:      es.Name,
+		},
+		Spec: mcsv1a1.ServiceImportSpec{
+			Type:  es.Service.Type,
+			Ports: es.Service.Ports,
+		},
+		Status: mcsv1a1.ServiceImportStatus{
+			Clusters: []mcsv1a1.ClusterStatus{{Cluster: se.ClusterName}},
+		},
+	}
+	if err != nil {
+		if apierrors.IsNotFound(err) {
+			err = r.Client.Create(ctx, &expected)
+			if err != nil {
+				log.Error(err, "unable to create the needed Service Import")
+			}
+			return err
+		}
+		log.Error(err, "Failed to get the needed service import")
+		return err
+	}
+	err = r.updateServiceImport(ctx, &si, &expected, log)
+	return err
+}
+
+// update the given ServiceImport to match the expected serviceImport
+func (r *ServiceExportReconciler) updateServiceImport(ctx context.Context, si, expected *mcsv1a1.ServiceImport, log logr.Logger) error {
+	si.Spec.Type = expected.Spec.Type
+	for i, expectedPort := range expected.Spec.Ports {
+		if i < len(si.Spec.Ports) {
+			expectedPort.DeepCopyInto(&si.Spec.Ports[i])
+		} else {
+			si.Spec.Ports = append(si.Spec.Ports, expectedPort)
+		}
+	}
+
+	err := r.Client.Update(ctx, si)
```

Comment:
is it wise to update all imports anyway even if no change? should we check first if they changed? 
@elevran is it done automatically by the client lib ?

Reply:
What about handling of update conflicts?